internal/graph/graph.go

```diff
@@ -109,8 +149,25 @@ func Referrers(ctx context.Context, target content.ReadOnlyGraphStorage, desc oc
 			if image.Subject == nil || !content.Equal(*image.Subject, desc) {
 				continue
 			}
-			node.ArtifactType = image.Config.MediaType
+			node.ArtifactType = image.ArtifactType
+			if node.ArtifactType == "" {
+				node.ArtifactType = image.Config.MediaType
+			}
 			node.Annotations = image.Annotations
+		case ocispec.MediaTypeImageIndex:
+			fetched, err := fetchBytes(ctx, target, node)
+			if err != nil {
+				return nil, err
+			}
+			var index ocispec.Index
+			if err := json.Unmarshal(fetched, &index); err != nil {
+				return nil, err
+			}
+			if index.Subject == nil || !content.Equal(*index.Subject, desc) {
+				continue
+			}
+			node.ArtifactType = index.ArtifactType
+			node.Annotations = index.Annotations
 		default:
 			continue
 		}
```

Comment:
Regarding to line 174:
```go
if node.ArtifactType != "" && (artifactType == "" || artifactType == node.ArtifactType)
```
Should it be
```go
if artifactType == "" || artifactType == node.ArtifactType
```
?

Reply:
It seems an index can have a `subject` without an `artifactType`.

Reply:
According to the spec, `artifactType` is used (not empty) `when the manifest is used for an artifact`. So if `artifactType` is not presented, an index is not an artifact and should not be returned as a referrer.

> if node.ArtifactType != "" && (artifactType == "" || artifactType == node.ArtifactType)

This makes sure that `node.ArtifactType == "" && artifactType == ""` does not pass.